iroh/src/util/fs.rs

```diff
@@ -225,10 +228,115 @@ fn path_content_info0(path: impl AsRef<Path>) -> anyhow::Result<PathContent> {
     Ok(PathContent { size, files })
 }
 
+/// Helper function that translates a key that was derived from a path on a filesystem back to that
+/// path.
+///
+/// If `prefix` exists, it will be stripped before converting back to a path
+/// If `root` exists, will add the root as a parent to the created path
+/// Removes any null byte that has been appened to the key
+pub fn key_to_path(
+    key: impl AsRef<[u8]>,
+    prefix: Option<String>,
+    root: Option<PathBuf>,
+) -> anyhow::Result<PathBuf> {
+    let mut key = key.as_ref();
+    if key.is_empty() {
+        return Ok(PathBuf::new());
+    }
+    // if the last element is the null byte, remove it
+    if b'\0' == key[key.len() - 1] {
+        key = &key[..key.len() - 1]
+    }
+
+    let key = if let Some(prefix) = prefix {
+        let prefix = prefix.into_bytes();
+        if prefix[..] == key[..prefix.len()] {
+            &key[prefix.len()..]
+        } else {
+            anyhow::bail!("key {:?} does not begin with prefix {:?}", key, prefix);
+        }
+    } else {
+        key
+    };
+
+    // convert to a path
+    let path = PathBuf::from(OsStr::from_bytes(key));
```

Comment:
okay, so this is obviously wrong because it only works for unix

I'm not sure the right thing to do for both the conversion btw path to bytes and bytes to path is, because I'm not sure I understand how to get from a &[u8] to &[u16] that is needed to for the windows version of OsString.

I also don't understand what the expectations are if we, for example, create a key from a path on one platform, but then convert that same key to a path on a different platform.

Can I get some advice? Is it worth this frustration? Is it easier/better to just fail if we don't have utf-8? Need some help!

Reply:
okay, this is all not very great, but for now we should simplify probably

the import steps should looks sth like this
- `canoncialize` the path (this removes things like `..` and symbolic links)
- iterate over each component with `.components()`
- for each component check that it is valid utf8 (yes this is a limitation for now, but utf16 or other crazy things are just too hard to deal with for now)
- escape any `/` with `\/`
- join all components into a `String` with `/` as the seperator

- on the out step replace `/` with the OS specific seperator
- unescape `\/` to `/`

Reply:
Something roughly similar exists in `util::fs::canonicalize_path`. It does not escape / though but just assumes that they don't exist and fails otherwise. This is what is used to create the path metadata for collections.

Reply:
okay, adapted from `canonicalize_path`:

There is now a `canonicalized_path_to_string` method that takes a path (that it expects to be canonicalized already) and turns it into a string

`canonicalize_path` is renamed to `relative_canonicalized_path_to_string`, and calls `canonicalized_path_to_string`, with the `must_be_relative` param set to `true`

These are a mouthful, but they are specific. TBH, I didn't even look at `canonicalize_path` closely before because I thought the function was just a util function to "canonicalize a path", not a function that turned a canonicalized path into a string.

I opted for keeping the restriction that these helper functions expect a canonicalized paths already, because I didn't want to complicate them by having them touch the actual file system or have to potentially add `async` functionality to helper util functions.